pkg/envoy/instance.go

```diff
@@ -0,0 +1,380 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package envoy
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"time"
+
+	envoyAdmin "github.com/envoyproxy/go-control-plane/envoy/admin/v2alpha"
+
+	"istio.io/pkg/log"
+)
+
+const (
+	defaultName = "envoy"
+)
+
+// Config for an Envoy Instance.
+type Config struct {
+	// Options provides the command-line options to be passed to Envoy.
+	Options Options
+
+	// Name of the envoy instance, used for logging only. If not provided, defaults to "envoy".
+	Name string
+
+	// BinaryPath the path to the Envoy binary.
+	BinaryPath string
+
+	// WorkingDir to be used when running Envoy. If not set, the current working directory is used.
+	WorkingDir string
+
+	// AdminPort specifies the administration port for the Envoy server. If not set, will
+	// be determined by parsing the Envoy bootstrap configuration file.
+	AdminPort uint32
+
+	// SkipBaseIDClose skips calling Close on the BaseID, if one was specified.
+	SkipBaseIDClose bool
+}
+
+// Waitable specifies a waitable operation.
+type Waitable interface {
```

Comment:
Is this just context.Context?

Reply:
Not exactly. This interface attempts to simplify the process of waiting by allowing you to wait indefinitely or for a given duration. There are 2 cases for waiting:

1. Waiting for the server to go live. 
2. Waiting for the server process to terminate. In this case, it returns the error returned by the process. This isn't possible by waiting for `ctx.Done`.